### ISSUE TITLE
Make Http Status configurable for ext auth service when enabled for gRPC services

### ISSUE THREAD

**Issue by ramaraochavali** [envoyproxy/envoy#11079]
when ext auth gRPC service returns` Permission Denied(7)`, Envoy returns `Http Status : 200, gRPC Status: 7` to client if the client is gRPC service and `Http Status - 403` to client if the client is Http Service. Technically this is correct but it causes some confusion for gRPC services.

In this case, since the actual gRPC service has not started processing the request yet(ext authz rejected before), I think it might fall under this category https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md and  gRPC clients should be able to handle non `200`  Http response as documented there?

Discussed with @dio on Slack - and we felt it may be reasonable to add a configuration to ext auth (`http_status_for_grpc_on_deny` or some thing similar) that can dictate this behaviour and send `403` in http status for gRPC services if configured. 

Creating this issue to continue that discussion and gather other opinions.

**Comment by ramaraochavali** [comment]
@mosdev FYI. For the OPA issue we are seeing. What do you think?

**Comment by ramaraochavali** [comment]
no stalebot

**Comment by ramaraochavali** [comment]
@dio can you mark this help wanted?

**Comment by ramaraochavali** [comment]
Thank you

**Comment by dio** [comment]
@ramaraochavali if we want to support this, we need to allow setting response headers status when `local_reply_data.is_grpc_` here: https://github.com/envoyproxy/envoy/blob/a536fd84e5926eca56ce29ddef20b49138f09f16/source/common/http/utility.cc#L516. 

Probably:

```c++
    response_headers->setStatus(flag_to_allow_overridding_http_status,
                                    ? response_headers->getStatusValue()
                                    : std::to_string(enumToInt(Code::OK)));

```

While setting it from the `ext_authz` perspective is already supported, but requires the external authorization server to response with `DeniedResponse`. 

```go
	return &envoy_service_auth_v3.CheckResponse{
		HttpResponse: &envoy_service_auth_v3.CheckResponse_DeniedResponse{
			DeniedResponse: &envoy_service_auth_v3.DeniedHttpResponse{
				Status: &envoy_type_v3.HttpStatus{
					Code: envoy_type_v3.StatusCode_Forbidden,
				},
			},
		},
		Status: &status.Status{
			Code: int32(code.Code_PERMISSION_DENIED),
		},
	}, nil
```

Or you can provide mapper:

```yaml
          local_reply_config:
            mappers:
            - filter:
                grpc_status_filter:
                  statuses:
                  - PERMISSION_DENIED
              status_code: 403
```

**Comment by ramaraochavali** [comment]
@dio  Thank you for digging in to this. I am thinking that based on some flag either in ext authz configuration or a generic flag (prefer ext authz config), add the following code

```
  response_headers->setStatus(flag_to_allow_overridding_http_status,
                                    ? response_headers->getStatusValue()
                                    : std::to_string(enumToInt(Code::OK)));
```
I think that would be lot cleaner than every on defining local mappers.

**Comment by htuch** [comment]
@markdroth is this a sensible thing to be doing from gRPC perspective?

**Comment by markdroth** [comment]
Thanks for tagging me.  I'm not familiar in depth with ext_authz, but at a high level, the stated intent here isn't entirely clear to me.

> when ext auth gRPC service returns` Permission Denied(7)`, Envoy returns `Http Status : 200, gRPC Status: 7` to client if the client is gRPC service and `Http Status - 403` to client if the client is Http Service. Technically this is correct but it causes some confusion for gRPC services.

What confusion does it cause?  If we actually know what gRPC status we want to return to the client, then sending that status directly sounds like exactly the right behavior.  Why convert that to an HTTP status just to make the client convert it back?

> In this case, since the actual gRPC service has not started processing the request yet(ext authz rejected before), I think it might fall under this category https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md and gRPC clients should be able to handle non `200` Http response as documented there?

As indicated at the top of that doc, the HTTP-to-gRPC status mapping is really intended to handle cases where an intermediate HTTP/2 proxy that is not gRPC-aware sends an HTTP error code without sending gRPC status.  The mapping allows gRPC to convert that to a gRPC status code.  But if the proxy *is* gRPC-aware, it should just send the gRPC status code directly, so that the client doesn't have to infer the mapping.

Note that the HTTP-to-gRPC status mapping is inherently lossy: there are a large number of gRPC status codes that cannot be expressed as an HTTP status code.  If, for example, the ext_authz server says to fail the RPC with status `INVALID_ARGUMENT`, converting it to an HTTP status code would make the client see `UNKNOWN` instead, which is basically lying to the client about what the real status code was.  That seems like strictly worse behavior than what you describe we currently have.

Now, all of that having been said, I do notice that the ext_authz service returns a [gRPC status code](https://github.com/envoyproxy/envoy/blob/6dc4092fc161b8fcb5f25a972e4783116692d53c/api/envoy/service/auth/v3/external_auth.proto#L114) but also an [HTTP status code to fail the request with](https://github.com/envoyproxy/envoy/blob/6dc4092fc161b8fcb5f25a972e4783116692d53c/api/envoy/service/auth/v3/external_auth.proto#L51).  That seems a little confusing; if both are set, what status is the client expected to use for the request?  Maybe the real problem here is that the semantics of the ext_authz server's response do not clearly indicate what status to use for gRPC requests?

**Comment by ramaraochavali** [comment]
>  That seems a little confusing; if both are set, what status is the client expected to use for the request? Maybe the real problem here is that the semantics of the ext_authz server's response do not clearly indicate what status to use for gRPC requests?

 This exactly the problem. Envoy returns Http Status : 200, gRPC Status: 7 to client if the client is gRPC service and Http Status - 403 to client if the client is Http Service.  We build common dashboards and alerts based on authz response. So it is little difficult to know (not impossible) whether we should look at http status code / gRPC status code. So this request is such that we can uniformly look at http status code in all cases. Definitely open to any other suggestions?

**Comment by markdroth** [comment]
In that case, how about adding a new field in the ext_authz server response that allows the server to explicitly indicate a gRPC status to use for the data plane request?  That way, there's no ambiguity: the server can say "if it's a gRPC call, fail it with this gRPC status; otherwise, fail it with this HTTP status."

**Comment by ramaraochavali** [comment]
And Envoy sets the same values to downstream clients?

**Comment by markdroth** [comment]
Yes, Envoy would fail the requests with whatever gRPC status the ext_authz server specifies.

**Comment by dclendenan** [comment]
@markdroth the issue is that failing with an HTTP 200/OK in the gRPC case, which doesn't provide the uniform indicator that @ramaraochavali described.  The gRPC status (DENIED/7) is already being passed through.  My read of the issue is that we want to have a non-OK HTTP status for gRPC calls which are rejected.  Would it fit to have a non-OK HTTP status accompany the gRPC status for ext_authz rejections?

cc: @dio

**Comment by markdroth** [comment]
If the grpc-status header is sent, then gRPC clients *should* ignore the HTTP status, so that should be fine in principle.  However, there may be some buggy clients that don't do this right, where the HTTP status might mask the gRPC status, so I do think it's important to have a way to send gRPC status without sending non-OK HTTP status.

Maybe the best thing would be to have one knob to specify the HTTP status and another knob to specify the gRPC status.  The ext_authz server can set either or both of the knobs, as appropriate for the environment.

**Comment by dclendenan** [comment]
In our use case it's OPA sending a gRPC status that is correct, but the HTTP 200/OK comes from Envoy and causes some confusion.  [My current PR](https://github.com/envoyproxy/envoy/pull/15979/files#diff-c22054b3e08df13d3a78842825935cd41aa61f439170a34e5f055e826da4d9c1R533) adds a switch that preserves the status from the HTTP headers of the response to the ext_authz call (OPA sends 403 in this case).    If the flag is set then both gRPC and HTTP response from Envoy reflect the ext-authz response, if not then gRPC from ext-authz is sent along but HTTP is 200/OK (the current behaviour).
I'm not quite clear where a new field in the ext_authz server response makes this smoother, if that then requires a change in the external services themselves (like OPA).
